File: Lua/day2.lua

function read_input2()
    lines = {}
    for line in io.lines('input2.txt') do
        lines[#lines + 1] = line
    end
    return lines
end


function Split(s, delimiter)
    result = {};
    for match in (s..delimiter):gmatch('(.-)'..delimiter) do
        table.insert(result, match);
    end
    return result;
end


function day2_1()
    lines = read_input2()
    depth = 0
    horizontal = 0
    for k, v in pairs(lines) do
        v = Split(v, ' ')
        direction = v[1]
        amount = tonumber(v[2])
        if direction == 'forward' then
            horizontal = horizontal + amount
        elseif direction == 'up' then
            depth = depth - amount
        elseif direction == 'down' then
            depth = depth + amount
        end
    end
    return depth * horizontal
end


function day2_2()
    lines = read_input2()
    depth = 0
    horizontal = 0
    aim = 0
    for k, v in pairs(lines) do
        v = Split(v, ' ')
        direction = v[1]
        amount = tonumber(v[2])
        if direction == 'forward' then
            horizontal = horizontal + amount
            depth = depth + aim * amount
        elseif direction == 'up' then
            aim = aim - amount
        elseif direction == 'down' then
            aim = aim + amount
        end
    end
    return depth * horizontal
end

File: Python/day2.py
from aocd import get_data
from utils import *

data = get_data(year=2021, day=2)


def part1():
    lines = splitlines(data, ' ')
    depth = 0
    horizontal = 0
    for direction, amount in lines:
        amount = int(amount)
        if direction == 'forward':
            horizontal += amount
        elif direction == 'up':
            depth -= amount
        elif direction == 'down':
            depth += amount
    return depth * horizontal


def part2():
    lines = splitlines(data, ' ')
    depth = 0
    horizontal = 0
    aim = 0
    for direction, amount in lines:
        amount = int(amount)
        if direction == 'forward':
            horizontal += amount
            depth += aim * amount
        elif direction == 'up':
            aim -= amount
        elif direction == 'down':
            aim += amount
    return depth * horizontal


if __name__ == '__main__':
    print(part1())
    print(part2())
